src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -1229,23 +1229,125 @@ private int SkipBlockComment(int i)
+                default: return c.ToString();
             }
         }
 
+        private string ScanUnicodeEscapeSequence()
+        {
+            // Unicode replacement char, used to represent unknown, unrecognized or unrepresentable characters.
+            var unknownChar = ((char)0xffdd).ToString();
+            int escSeqStartIndex = _currentIndex - 2;
+            bool isBracketedSequence = false;
+            bool isTerminated = false;
+
+            char c = PeekChar();
+            if (c == '{')
+            {
+                SkipChar();
+                isBracketedSequence = true;
+            }
+
+            int maxNumberOfHexDigits = (isBracketedSequence ? 6 : 4);
+
+            // Scan hex chars after the Unicode escape sequence start.
+            var sb = GetStringBuilder();
+            for (int i = 0; i < maxNumberOfHexDigits; i++)
+            {
+                c = PeekChar();
+
+                // Bracketed sequence has been terminated.
+                if (isBracketedSequence && (c == '}'))
+                {
+                    SkipChar();
+
+                    if (i == 0)
+                    {
+                        // Variable sequence must have at least one hex char.
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidBracketedUnicodeEscapeSequence);
+                        return unknownChar;
+                    }
+
+                    isTerminated = true;
+                    break;
+                }
+
+                if ((c == '\0') || !c.IsHexDigit())
+                {
+                    if (isBracketedSequence)
+                    {
+                        ReportError(_currentIndex, () => ParserStrings.UnicodeEscapeSequenceMissingTerminator);
+                    }
+                    else
+                    {
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                    }
+
+                    return unknownChar;
+                }
+
+                SkipChar();
+                sb.Append(c);
+            }
+
+            if (isBracketedSequence && !isTerminated)
+            {
+                c = PeekChar();
+                if (c != '}')
+                {
+                    ReportError(_currentIndex, () => ParserStrings.UnicodeEscapeSequenceMissingTerminator);
+                    return unknownChar;
+                }
+
+                SkipChar();
+            }
+
+            string hexStr = GetStringAndRelease(sb);
+
+            int unicodeValue;
+            if (!int.TryParse(hexStr, NumberStyles.AllowHexSpecifier, NumberFormatInfo.InvariantInfo, out unicodeValue))
+            {
+                // Considering we validate hex characters, we should never get here.
```

Comment:
Thanks! I was wondering about that. Will remove the corresponding ParserStrings entry.